test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,370 @@
+  ON_CALL(*cluster2_info, statsScope()).WillByDefault(ReturnRef(cluster2_stats_scope));
+
+  // set gauge value.
+  ON_CALL(cluster2_stats_scope, gauge("membership_total"))
+      .WillByDefault(ReturnRef(membership_total_gauge_));
+
+  // attach counters
+  NiceMock<Stats::MockCounter> success_counter2;
+  NiceMock<Stats::MockCounter> error_5xx_counter2;
+  NiceMock<Stats::MockCounter> retry_5xx_counter2;
+  NiceMock<Stats::MockCounter> error_4xx_counter2;
+  NiceMock<Stats::MockCounter> retry_4xx_counter2;
+  success_counter2.name_ = "upstream_rq_2xx";
+  ON_CALL(cluster2_stats_scope, counter("upstream_rq_2xx"))
+      .WillByDefault(ReturnRef(success_counter2));
+  error_5xx_counter2.name_ = "upstream_rq_5xx";
+  ON_CALL(cluster2_stats_scope, counter("upstream_rq_5xx"))
+      .WillByDefault(ReturnRef(error_5xx_counter2));
+  retry_5xx_counter2.name_ = "retry.upstream_rq_5xx";
+  ON_CALL(cluster2_stats_scope, counter("retry.upstream_rq_5xx"))
+      .WillByDefault(ReturnRef(retry_5xx_counter2));
+  error_4xx_counter2.name_ = "upstream_rq_4xx";
+  ON_CALL(cluster2_stats_scope, counter("upstream_rq_4xx"))
+      .WillByDefault(ReturnRef(error_4xx_counter2));
+  retry_4xx_counter2.name_ = "retry.upstream_rq_4xx";
+  ON_CALL(cluster2_stats_scope, counter("retry.upstream_rq_4xx"))
+      .WillByDefault(ReturnRef(retry_4xx_counter2));
+
+  cluster2_info->stats().upstream_rq_timeout_.reset();
+  cluster2_info->stats().upstream_rq_per_try_timeout_.reset();
+  cluster2_info->stats().upstream_rq_pending_overflow_.reset();
+
+  // add cluster
+  cluster_map_.emplace(cluster2_name, cluster2);
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  // generate data to both clusters
+  for (int i = 0; i < 12; i++) {
+    buffer.drain(buffer.length());
+    // cluster 1
+    ON_CALL(error_5xx_counter_, value()).WillByDefault(Return((i + 1) * 17));
+    ON_CALL(success_counter_, value()).WillByDefault(Return((i + 1) * 7));
+    cluster1_info_->stats().upstream_rq_timeout_.add(1);
+    cluster1_info_->stats().upstream_rq_per_try_timeout_.add(2);
+    cluster1_info_->stats().upstream_rq_pending_overflow_.add(8);
+
+    // cluster 2
+    ON_CALL(error_5xx_counter2, value()).WillByDefault(Return((i + 1) * 1));
+    ON_CALL(retry_5xx_counter2, value()).WillByDefault(Return((i + 1) * 2));
+    ON_CALL(error_4xx_counter2, value()).WillByDefault(Return((i + 1) * 3));
+    ON_CALL(retry_4xx_counter2, value()).WillByDefault(Return((i + 1) * 4));
+    ON_CALL(success_counter2, value()).WillByDefault(Return((i + 1) * 3));
+    cluster2_info->stats().upstream_rq_timeout_.add(3);
+    cluster2_info->stats().upstream_rq_per_try_timeout_.add(1);
+    cluster2_info->stats().upstream_rq_pending_overflow_.add(5);
+
+    sink_->flush(source_);
+  }
+
+  std::string data_message = TestUtility::bufferToString(buffer);
+
+  std::size_t pos1 = data_message.find(cluster1_name_);
+  std::size_t pos2 = data_message.find(cluster2_name);
```

Comment:
I wonder if this pattern of using find on the cluster-names and this dance with the substr etc could be done easier with absl::StrSplit. That would be the case if there were a known delimiter between the clusters.

Reply:
A helper method to populate & return a map would be be useful to remove this kind of complicated pattern of finding substring delimiters in this way. In addition to require a handful of lines that require some scrutiny, the current pattern could fail if either cluster-name turned out to be a substring of some other syntax in the data.